examples/roam/src/components/Block.js

```diff
@@ -0,0 +1,383 @@
+      ) : null}
+    </div>
+  )
+}
+
+const BlockEditString = ({ id: inputId }) => {
+  const [block] = useEntity(inputId)
+  const [transact] = useTransact()
+  const [pageBlocks] = useQuery({
+    $find: 'block',
+    $where: { block: { title: '$any' } },
+  })
+  const titles = React.useMemo(() => pageBlocks.map((b) => b.get('title') || []), [pageBlocks])
+  const id = block.get('id') || inputId
+  // Use react component level state for added performance
+  // since this has to be a controlled textarea component
+  // for ReactTextareaAutocomplete to work
+  const [textareaString, setTextareaString] = React.useState(null)
+  // Put this transaction in a callback so it can be debounced
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const setBlockString = React.useCallback(
+    debounce((str) => {
+      transact([{ block: { id, string: str } }])
+    }, 300),
+    [transact, id],
+  )
+  const rtaRef = React.useRef()
+  React.useEffect(() => {
+    const i = block.get('editingStartingCaretIndex') || 99999
+    rtaRef.current.setCaretPosition(i === 1 ? 99999 : i)
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [])
+
+  return (
+    <ReactTextareaAutocomplete
+      autoFocus
+      minChar={0}
+      renderToBody
+      ref={rtaRef}
+      className="w-full outline-none bg-transparent"
+      containerClassName="text-base flex"
+      onChange={(e) => {
+        // This instantly updates the UI
+        setTextareaString(e.target.value || '')
+        // This saves the string to homebase and ultimately
+        // the backend API (Firebase) asynchronously to avoid slowing the UI
+        setBlockString(e.target.value)
+      }}
+      onBlur={() => {
+        transact([
+          {
+            block: {
+              id,
+              isEditing: null,
+              editingStartingCaretIndex: null,
+            },
+          },
+        ])
+        setTextareaString(null)
+      }}
+      value={textareaString !== null ? textareaString : block.get('string') || ''}
+      textAreaComponent={{ component: TextareaAutosize, ref: 'innerRef' }}
+      loadingComponent={() => <span>Loading</span>}
+      trigger={{
+        '[[': {
+          dataProvider: (token) => {
+            const re = new RegExp(token.slice(1), 'i')
+            return [...titles.filter((title) => re.test(title)).slice(0, 9), token.slice(1)]
+          },
+          component: ({ entity }) => <div className="py-1 px-2">{entity}</div>,
+          output: (entity) => ({ text: `[[${entity}]]`, caretPosition: 'next' }),
+        },
+      }}
+      onKeyDown={(e) => {
+        if (e.key === 'Enter' && !e.shiftKey) {
+          e.preventDefault()
+          const parentId = block.get('_children', 0, 'id')
+          const order = block.get('order') || 0
+          const siblingOrders = block.get('_children', 0, 'children', 'order')
+          const nextSiblingOrder = siblingOrders.filter((v) => v > order).sort()[0] || order + 2
```

Comment:
it's unclear here what the constants (2, 2.0)  are doing. Maybe we can abstract and make clearer?

Reply:
This is tricky since `2` is a bit arbitrary and this ordering code, in particular, is pretty hacky. 

I'll rename the var to see if that clarifies anything.